agent-control/src/agent_control/config.rs

```diff
@@ -65,82 +65,148 @@ pub struct AgentControlConfig {
+            let mut yaml_configuration =
+                YAMLConfig::try_from(config_content.as_str()).map_err(|err| {
+                    AgentControlConfigError(format!("invalid config '{}': {}", config_name, err))
+                })?;
+
+            if let Some(agents_value) = yaml_configuration.remove_key(AGENTS_KEY) {
+                merged_agents = try_append_agents(merged_agents, agents_value).map_err(|err| {
+                    AgentControlConfigError(format!(
+                        "appending agents from config '{}': {}",
+                        config_name, err
+                    ))
+                })?;
+            }
 
-impl TryFrom<YAMLConfig> for AgentControlConfig {
-    type Error = serde_yaml::Error;
+            remaining_config = YAMLConfig::try_append(remaining_config.clone(), yaml_configuration)
+                .map_err(|err| AgentControlConfigError(format!("appending config: {err}")))?;
+        }
 
-    fn try_from(value: YAMLConfig) -> Result<Self, Self::Error> {
-        serde_yaml::from_value(serde_yaml::to_value(value)?)
+        let remaining_config_appended = AgentControlDynamicConfig::try_from(remaining_config)
+            .map_err(|err| AgentControlConfigError(format!("encoding config: {}", err)))?;
+
+        Ok(AgentControlDynamicConfig {
+            agents: merged_agents,
+            ..remaining_config_appended
+        })
     }
 }
 
-impl TryFrom<&AgentControlDynamicConfig> for YAMLConfig {
-    type Error = serde_yaml::Error;
-
-    fn try_from(value: &AgentControlDynamicConfig) -> Result<Self, Self::Error> {
-        serde_yaml::from_value(serde_yaml::to_value(value)?)
+/// Tries to append agents from a YAML value into the agents map, erroring on duplicates.
+fn try_append_agents(
+    merged_agents: SubAgentsMap,
+    agents_value: serde_yaml::Value,
+) -> Result<SubAgentsMap, AgentControlConfigError> {
+    let sub_agents_map: SubAgentsMap = serde_yaml::from_value(agents_value)
+        .map_err(|err| AgentControlConfigError(format!("invalid agents: {}", err)))?;
+
+    let mut merged_agents = merged_agents;
+
+    for (agent_id, agent_config) in sub_agents_map {
+        if merged_agents
+            .insert(agent_id.clone(), agent_config)
+            .is_some()
+        {
+            return Err(AgentControlConfigError(format!(
+                "duplicated agent: {}",
+                agent_id
+            )));
+        }
     }
+
+    Ok(merged_agents)
 }
 
 impl TryFrom<YAMLConfig> for AgentControlDynamicConfig {
-    type Error = serde_yaml::Error;
+    type Error = AgentControlConfigError;
 
     fn try_from(value: YAMLConfig) -> Result<Self, Self::Error> {
-        serde_yaml::from_value(serde_yaml::to_value(value)?)
+        serde_yaml::from_value(
+            serde_yaml::to_value(value)
+                .map_err(|e| AgentControlConfigError(format!("deserializing: {e}")))?,
+        )
+        .map_err(|e| AgentControlConfigError(format!("serializing: {e}")))
```

Comment:
Didn't remember this `T -> Value -> U` conversion. I'd say it works well enough for us! 👌